### ISSUE TITLE
OAuth2 login error due to wrong parsing of JWT

### ISSUE THREAD

**Issue by trybowski** [roundcube/roundcubemail#8890]
There is an error in rcmail_oauth.php that prevents login in certain cases. The reason is that JWT token can contain characters such as: underscore (`_`) and dash (`-`) and these are not accepted by base64_decode.

Please note that JWT's are not base64-encoded, they are base64url-encoded. The difference is that where base64 uses slash (/) and plus (+), base64url uses underscore (`_`) and dash (`-`). Therefore the JWT must be first translated from base64url to base64 and only then it can be treated by base64_decode.

To to this, all `-` must be replaced with `+` and all `_` with `/`. The code could be: `$jwt = strtr($jwt, '-_', '+/');`

File program/include/rcmail_oauth.php needs to be changed in line 145:

    public function jwt_decode($jwt)
    {
        list($headb64, $bodyb64, $cryptob64) = explode('.', $jwt);

to

    public function jwt_decode($jwt)
    {
        $jwt = strtr($jwt, '-_', '+/');
        list($headb64, $bodyb64, $cryptob64) = explode('.', $jwt);

Further reading:
- https://stackoverflow.com/questions/49082844/how-could-firebase-send-a-jwt-token-which-payload-contains-an-underscore-charact
- https://www.base64url.com/
- https://www.php.net/manual/en/function.base64-decode.php

**Comment by thomascube** [comment]
Fixed for 1.6.2

**Comment by LauraTaylorUK** [comment]
I am not sure if its relevant here, but the base64url-encode/decode also require the removal of `=` (encode) and padding with `=` (decode). The JWT library uses these functions:

```php
    /**
     * Decode a string with URL-safe Base64.
     *
     * @param string $input A Base64 encoded string
     *
     * @return string A decoded string
     */
    public static function urlsafeB64Decode($input)
    {
        $remainder = \strlen($input) % 4;
        if ($remainder) {
            $padlen = 4 - $remainder;
            $input .= \str_repeat('=', $padlen);
        }
        return \base64_decode(\strtr($input, '-_', '+/'));
    }

    /**
     * Encode a string with URL-safe Base64.
     *
     * @param string $input The string you want encoded
     *
     * @return string The base64 encode of what you passed in
     */
    public static function urlsafeB64Encode($input)
    {
        return \str_replace('=', '', \strtr(\base64_encode($input), '+/', '-_'));
    }
```
